File: dotnet/videostore/Statement.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace videostore
{
    class Statement
    {
        public int FrequentRenterPoints { get; set; }
        public double AmountOwed { get; set; }
        protected string Name { get; private set; }
        private readonly IList<Rental> rentals;

        public Statement(String name)
        {
            Name = name;
            rentals = new List<Rental>();
        }

        public void AddRental(Rental rental)
        {
            rentals.Add(rental);
        }

        public String Generate()
        {
            AmountOwed = 0;
            FrequentRenterPoints = 0;

            var result = "Rental Record for " + Name + "\n";
            
            foreach (var rental in rentals)
            {
                var thisAmount = Rental.AmountFor(rental);

                FrequentRenterPoints++;

                if (rental.GetMovie().GetPriceCode() == Movie.NEW_RELEASE
                        && rental.GetDaysRented() > 1)
                    FrequentRenterPoints++;

                result += "\t" + rental.GetMovie().GetTitle() + "\t" + string.Format("{0:F1}", thisAmount) + "\n";
                AmountOwed += thisAmount;
            }

            result += "You owed " + string.Format("{0:F1}", AmountOwed) + "\n";
            result += "You earned " + FrequentRenterPoints + " frequent renter points\n";
            return result;
        }
    }
}

File: java/videostore/src/Statement.java
import java.util.ArrayList;
import java.util.List;

class Statement {

  private final String customerName;
  private final List<Rental> rentals = new ArrayList<Rental>();
  private double totalAmount;
  private int frequentRenterPoints;

  Statement(String customerName) {
    this.customerName = customerName;
  }

  void addRental(Rental rental) {
    rentals.add(rental);
  }

  String generate() {
    clearTotals();
    String statementText = header();
    statementText += rentalLines();
    statementText += footer();
    return statementText;
  }

  private void clearTotals() {
    totalAmount = 0;
    frequentRenterPoints = 0;
  }

  private String header() {
    return String.format("Rental Record for %s\n", customerName);
  }

  private String rentalLines() {
    String rentalLines = "";
    for (Rental rental : rentals) {
      rentalLines += rentalLine(rental);
    }
    return rentalLines;
  }

  private String rentalLine(Rental rental) {
    double rentalAmount = rental.determineAmount();
    frequentRenterPoints += rental.determineFrequentRenterPoints();
    totalAmount += rentalAmount;
    return formatRentalLine(rental, rentalAmount);
  }

  private String formatRentalLine(Rental rental, double rentalAmount) {
    return String.format("\t%s\t%.1f\n", rental.getTitle(), rentalAmount);
  }

  private String footer() {
    return String.format(
        "You owed %.1f\n" + "You earned %d frequent renter points\n",
        totalAmount, frequentRenterPoints);
  }

  double getTotal() {
    return totalAmount;
  }

  int getFrequentRenterPoints() {
    return frequentRenterPoints;
  }
}
